File: v2/assets/mcp/hil-skill.py
#!/usr/bin/env python3
"""
ClawEnv HIL Skill — Human-in-the-Loop browser intervention via MCP (Python version)

Runs inside the sandbox as an MCP stdio server.
When an agent encounters a CAPTCHA, 2FA, or other browser challenge that
requires human intervention, it calls the `hil_request` tool.

Flow:
  1. Agent calls hil_request(reason)
  2. This skill POSTs to ClawEnv Bridge /api/hil/request
  3. ClawEnv switches browser to interactive (noVNC) mode
  4. User resolves the challenge and clicks "Continue Auto"
  5. ClawEnv responds to the HTTP request
  6. This skill returns success to the agent
  7. Agent continues automation

Register with Hermes Agent:
  hermes mcp add clawenv-hil --config '{"command":"python3","args":["/workspace/hil-skill/skill.py"]}'
"""
from __future__ import annotations

import sys
import os
import json
import argparse
import asyncio

import httpx
from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import TextContent

# Parse CLI args
parser = argparse.ArgumentParser()
parser.add_argument("--bridge-url", default="")
args = parser.parse_args()

bridge_url = args.bridge_url

# Auto-detect bridge URL
if not bridge_url:
    host_ip = os.environ.get("CLAWENV_HOST_IP")
    if host_ip:
        bridge_url = f"http://{host_ip}:3100"
    else:
        try:
            r = httpx.get("http://host.containers.internal:3100/api/health", timeout=2)
            if r.status_code == 200:
                bridge_url = "http://host.containers.internal:3100"
        except Exception:
            pass
        if not bridge_url:
            bridge_url = "http://host.lima.internal:3100"

print(f"[clawenv-hil] Bridge URL: {bridge_url}", file=sys.stderr)

# HTTP client with 10-min timeout for HIL blocking requests
client = httpx.AsyncClient(base_url=bridge_url, timeout=600)


def text(s: str) -> list:
    """Wrap a string in MCP TextContent list for tool responses."""
    return [TextContent(type="text", text=s)]


server = Server("clawenv-hil")


@server.tool()
async def hil_request(reason: str, url: str = "") -> list:
    """Request human intervention for a browser task.

    Use this when the browser encounters a CAPTCHA, 2FA prompt, login wall,
    cookie consent, or any challenge that cannot be automated.
    This will open the browser in interactive mode for the user.
    The call blocks until the user finishes and clicks 'Continue Auto'.
    """
    print(f"[clawenv-hil] HIL requested: {reason}{f' at {url}' if url else ''}", file=sys.stderr)

    try:
        resp = await client.post(
            "/api/hil/request",
            json={"reason": reason, "url": url},
        )

        if resp.status_code != 200:
            return text(f"HIL request failed: {resp.text}")

        data = resp.json()
        print(f"[clawenv-hil] HIL completed: {json.dumps(data)}", file=sys.stderr)

        result = f"Human intervention completed. {data.get('message', 'User finished the task.')}"
        if data.get("notes"):
            result += f"\nUser notes: {data['notes']}"
        return text(result)

    except Exception as e:
        return text(f"HIL request error: {e}")


@server.tool()
async def hil_status() -> list:
    """Check if a human intervention session is currently active."""
    try:
        resp = await client.get("/api/hil/status", timeout=5)
        data = resp.json()
        return text(json.dumps(data, indent=2))
    except Exception as e:
        return text(f"Cannot check HIL status: {e}")


async def main():
    async with stdio_server() as (read_stream, write_stream):
        print("[clawenv-hil] HIL skill server started on stdio", file=sys.stderr)
        await server.run(read_stream, write_stream, server.create_initialization_options())


if __name__ == "__main__":
    asyncio.run(main())

File: v2/assets/mcp/hil-skill.mjs
#!/usr/bin/env node
/**
 * ClawEnv HIL Skill — Human-in-the-Loop browser intervention via MCP
 *
 * Runs inside the sandbox as an MCP stdio server.
 * When an agent encounters a CAPTCHA, 2FA, or other browser challenge that
 * requires human intervention, it calls the `hil_request` tool.
 *
 * Flow:
 *   1. Agent calls hil_request(reason)
 *   2. This skill POSTs to ClawEnv Bridge /api/hil/request
 *   3. ClawEnv switches browser to interactive (noVNC) mode
 *   4. User resolves the challenge and clicks "Continue Auto"
 *   5. ClawEnv responds to the HTTP request
 *   6. This skill returns success to the agent
 *   7. Agent continues automation
 *
 * Register with OpenClaw:
 *   openclaw mcp set clawenv-hil '{"command":"node","args":["/workspace/hil-skill/index.mjs"]}'
 */
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { z } from "zod";

// Parse CLI args
const args = process.argv.slice(2);
let bridgeUrl = "";
for (let i = 0; i < args.length; i++) {
  if (args[i] === "--bridge-url" && args[i + 1]) {
    bridgeUrl = args[i + 1]; i++;
  }
}

// Auto-detect bridge URL
if (!bridgeUrl) {
  const hostIp = process.env.CLAWENV_HOST_IP;
  if (hostIp) {
    bridgeUrl = `http://${hostIp}:3100`;
  } else {
    try {
      const resp = await fetch("http://host.containers.internal:3100/api/health", { signal: AbortSignal.timeout(2000) });
      if (resp.ok) bridgeUrl = "http://host.containers.internal:3100";
    } catch { /* not Podman */ }
    if (!bridgeUrl) bridgeUrl = "http://host.lima.internal:3100";
  }
}

console.error(`[clawenv-hil] Bridge URL: ${bridgeUrl}`);

const server = new McpServer({
  name: "clawenv-hil",
  version: "0.1.0",
});

// === Tool: hil_request ===
// Blocks until human completes the intervention
server.tool(
  "hil_request",
  "Request human intervention for a browser task. " +
  "Use this when the browser encounters a CAPTCHA, 2FA prompt, login wall, " +
  "cookie consent, or any challenge that cannot be automated. " +
  "This will open the browser in interactive mode for the user. " +
  "The call blocks until the user finishes and clicks 'Continue Auto'.",
  {
    reason: z.string().describe(
      "Why human help is needed (e.g. 'CAPTCHA detected on login page', '2FA code required')"
    ),
    url: z.string().optional().describe(
      "The URL where intervention is needed (informational, shown to user)"
    ),
  },
  async ({ reason, url }) => {
    console.error(`[clawenv-hil] HIL requested: ${reason}${url ? ` at ${url}` : ""}`);

    try {
      // This POST blocks until the user completes intervention
      const resp = await fetch(`${bridgeUrl}/api/hil/request`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ reason, url: url ?? "" }),
        signal: AbortSignal.timeout(600_000), // 10 min max wait
      });

      if (!resp.ok) {
        const text = await resp.text();
        return {
          content: [{ type: "text", text: `HIL request failed: ${text}` }],
          isError: true,
        };
      }

      const data = await resp.json();
      console.error(`[clawenv-hil] HIL completed: ${JSON.stringify(data)}`);

      return {
        content: [{
          type: "text",
          text: `Human intervention completed. ${data.message ?? "User finished the task."}` +
                `${data.notes ? `\nUser notes: ${data.notes}` : ""}`,
        }],
      };
    } catch (err) {
      return {
        content: [{ type: "text", text: `HIL request error: ${err.message}` }],
        isError: true,
      };
    }
  }
);

// === Tool: hil_status ===
// Check if HIL is currently active
server.tool(
  "hil_status",
  "Check if a human intervention session is currently active.",
  {},
  async () => {
    try {
      const resp = await fetch(`${bridgeUrl}/api/hil/status`, { signal: AbortSignal.timeout(5000) });
      const data = await resp.json();
      return {
        content: [{ type: "text", text: JSON.stringify(data, null, 2) }],
      };
    } catch (err) {
      return {
        content: [{ type: "text", text: `Cannot check HIL status: ${err.message}` }],
        isError: true,
      };
    }
  }
);

const transport = new StdioServerTransport();
await server.connect(transport);
console.error("[clawenv-hil] HIL skill server started on stdio");
